iocore/net/SSLCertLookup.cc

```diff
@@ -47,24 +48,18 @@
 struct SSLAddressLookupKey {
   explicit SSLAddressLookupKey(const IpEndpoint &ip) : sep(0)
   {
-    static const char hextab[16] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};
-
-    int nbytes;
-    uint16_t port = ntohs(ip.port());
-
     // For IP addresses, the cache key is the hex address with the port concatenated. This makes the lookup
     // insensitive to address formatting and also allow the longest match semantic to produce different matches
     // if there is a certificate on the port.
-    nbytes = ats_ip_to_hex(&ip.sa, key, sizeof(key));
+
+    ts::FixedBufferWriter w{key, sizeof(key)};
+    w.print("{:x}", std::string_view{reinterpret_cast<char const *>(ats_ip_addr8_cast(&ip)), ats_ip_addr_size(&ip)});
```

Comment:
This requires much knowledge. Why don't we have a function like `size_t to_hex_str(char *out, const char *in, size_t in_len)` ?

Reply:
I disagree. This uses a much more general mechanism, where the format code 'x' (or 'X' for upper case) means dump a string view as hex. The `BufferWriter` embeds the pointer and length in a safe way that is easily used. In this case, for instance, the developer doesn't have to know multiple things (one for the IP address, one for the port, as the code is now) but only one thing, 'x' means hex dump for a string view. Note also `BufferWriter` is safer than your suggestion, because it prevents buffer overruns, rather than hoping the output buffer is long enough. It's also frequently more compact - in this case, if the port were always added (which would seem to be fine), then it could be done in one line, rather than multiple calls to various different functions.

This could definitely be improved. If I had a free hand, I would change the formatting for IP addresses such that a format code of "x" would mean dump as hex. I would also enable the `FixedBufferWriter` constructor to take a `char` array directly. It would also be possible to add a general format type, such as  say 'm', to mean "dump as hex" for any type.

Reply:
I'm saying we need to know the name of FixedBufferWriter, format codes, print function and write function. `to_hex_str` is more easy to use, isn't it? You can see how to use the function without references. Your comment sounds like creating wrappers of printf is not a good idea.

If `to_hex_str` have out, out_len, in and in_len (I added out_len), we can do the same thing inside the function with FixedBufferWriter, no?

Reply:
The string format notation of FixedBufferWriter is the same as the python format strings, so it is not so out of the blue.

In general, I think this tidy up is good.  Getting rid of explicit hex table manipulations.

Reply:
I don't think `to_hex_str` is easier to use _in general_, and not even in this specific case (although that's debatable), because if there is other formatting happening, such as detecting the value for `sep`, and adding the separating '.', and putting that all after the IP address. Note with BW there is no need to track how much space `to_hex_str` used, BW does that accounting for me, along with preventing buffer overruns.

I think it is nice to have one mechanism that does all of the output formatting. In essence, I did write `to_hex_str` and then tweaked BW to call it if the 'x' format code is used on a `string_view`, allowing it to intermix with other formatting very easily. If used generally, I think this will definitely be easier, but getting started is challenging. That's why I wanted to make this PR, to provide an example.

Also, I updated the PR to remove the `write`, so now there is just `print`.

Reply:
To get a taste of where I would like to go with this, and how it cleans up existing code, see [here](https://github.com/SolidWallOfCode/trafficserver/commit/5e14faea94bf4b94d5258040cd6d97f567f8bf5e) for cleaning up the diagnostic logging code, which was inspired by your work on the thread name support.

Reply:
> In general, I think this tidy up is good. Getting rid of explicit hex table manipulations.

Well, maybe, I'm not too sure.

A reason there are many hex converter I think is that we don't have a standard way to do it. There would be special cases (and here may be one of the cases), but I think people would not write own hex converter if we had to_hex_str in tscore. This is what I expect to #4637. However, this change looks like just doing the conversion in a different way. (Using FixedBufferWriter is fine. The point is this change is just using a generic way.)

People would copy and paste these lines if they could find it, but I think finding conversion code would get difficult because there is no "hex" in these lines. And it produces more hex converters, IMO.

Also, if somebody introduced a better way, we would need to update all these hex conversion that use FixedBufferWriter directly.

I can live with this change, but I don't think we can close #4637 even if we replace all the hex table manipulations with FixedBufferWriter.

Reply:
Why not? It would unify all of the hex conversions in one place. My goal here is to make using `BufferWriter` the standard way, because I think overall that's the best option. That is, if you want to format output in hex, you use the 'x' format code, just the same way as you would for printf, except now it applies to _all_ types, not just integers. That's the main reason I added the hex dump capability to printing `string_view`.

Another alternative here is to use a wrapper class, like `bwf::Errno` or `bwf::Date`. It could be `bwf::AsHex` or `bwf::HexDump` and then you would do
```
w.print("{}", bwf::AsHex(ip));
```
That would be extremely simple to implement, if you'd like that better.

Reply:
Because BufferWriter is too generic and the interface doesn't say "This is the one you should use when you want to print data in hex.". How people can know BufferWriter is the one we provide as a standard way to print data in hex? It looks like one of options to me. And as I mentioned above, lack of explicit hex converter would make people write own converters. Some may use BufferWriter, but some may not. Then we need to reopen the issue to unify them again. Of course we can point it out on PRs, but how would you feel if you were told "use this instead of that" after you wrote an function because you couldn't find the existing function that has the same feature due to the naming?

>  It could be bwf::AsHex or bwf::HexDump

These look better than current because these can be find with "grep -i hex", but that is not my point exactly.

Reply:
Alright, let's see how it goes then.

My understanding is that making shorthands for specific purposes is a bad idea and we should use BufferWriter without any wrappers because BufferWriter can cover any sort of conversions with the best interface.

Reply:
Essentially yes - let's have one generic mechanism rather than a grab bag of context dependent functions. I think in the end this will be simpler for developers.